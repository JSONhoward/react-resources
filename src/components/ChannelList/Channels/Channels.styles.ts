import styled from 'styled-components';

export const StyledChannels = styled('div')`
display: flex;
flex-direction: row;
align-items: center;
height: 250px;
width: 90%;
box-shadow: 1px 1px 15px rgba(0,0,0,.25);
border-radius: 5px;
margin: 2rem;

@media screen and (max-width: 800px) {
    height: 200px;
}

@media screen and (max-width: 600px) {
    height: 100px;
}
`

export const ChannelsLoading = styled('div')`
display: flex;
flex-direction: row;
align-items: center;
height: 250px;
width: 90%;
background: ${({theme: {ReactSecondary}}) => ReactSecondary};
box-shadow: 1px 1px 15px rgba(0,0,0,.25);
border-radius: 5px;
margin: 2rem;
`

export const ChannelIcon = styled('div')`
display: flex;
flex-direction: column;
align-items: center;
justify-content: space-evenly;
height: 100%;
width: 250px;
overflow: hidden;
margin: 0 0rem;

img {
height: 75%;
width: 75%;
object-fit: cover;
border-radius: 50%;
box-shadow: 1px 1px 5px rgba(0,0,0,.25);
cursor: pointer;
}

@media screen and (max-width: 800px) {
    width: 200px;
}

@media screen and (max-width: 600px) {
    width: 100px;
}
`

export const ChannelName = styled('p')`
text-align: center;

@media screen and (max-width: 600px) {
    display: none;
}
`

export const ChannelVideos = styled('div')`
position: relative;
display: flex;
align-items: center;
justify-content: space-evenly;
height: 100%;
width: 100%;
border-top-right-radius: 5px;
border-bottom-right-radius: 5px;
background: ${({theme: {ReactSecondary}}) => ReactSecondary};
padding: .5rem;

a, a:active, a:visited {
    text-decoration: none;
    color: ${({theme: {ReactDark}}) => ReactDark};
}

a:hover {
color: ${({theme: {ReactPrimary}}) => ReactPrimary};
}
`

export const VideoImagePlay = styled('div')`
position: absolute;
height: 100%;
width: 100%;
backdrop-filter: blur(2px);
overflow: hidden;
opacity: 0;
transition: all .5s ease-in-out;
cursor: pointer;

&:hover {
    opacity: 1;
}

svg {
    position: absolute;
    top: 50%;
    left: 50%;
    transform: translate(-50%, -50%);
}

span {
    position: absolute;
    top: .5rem;
    left: 50%;
    transform: translateX(-50%);
    color: white;
    text-shadow: 1px 1px 5px black;
}
`

export const VideoImage = styled('div')`
position: relative;
height: 100%;
width: 33%;
margin: .5rem;

img {
    height: 100%;
    width: 100%;
    object-fit: cover;
}
`

export const More = styled('div')`
position: absolute;
top: 50%;
right: 0;
display: grid;
place-items: center;
height: 50%;
width: 4.5rem;
transform: translateY(-50%);
border-top-left-radius: 50px;
border-bottom-left-radius: 50px;
background: white;
cursor: pointer;
overflow: hidden;
transition: width 1s ease-in-out, color 1s ease-in-out;

p {
    text-align: center;
    width: 10rem;
}

&:hover {
width: 9rem;
}
`