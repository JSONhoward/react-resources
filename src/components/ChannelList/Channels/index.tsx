import ChannelList from './Channels';
export default ChannelList