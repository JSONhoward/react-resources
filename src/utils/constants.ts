import cleverP from '../images/clever-programmer-icon.jpg'
import devEd from '../images/dev-ed-icon.jpg'
import fireship from '../images/fireship-icon.jpg'
import freeCodeCamp from '../images/free-code-camp-icon.jpg'
import netNinja from '../images/net-ninja-icon.jpg'
import webDevSimplified from '../images/web-dev-simplified-icon.jpg'
import rwieruch from '../images/rwieruch-icon.svg'
import sitePoint from '../images/sitepoint.svg'
import smashingMagazine from '../images/smashing-magazine.svg'
import codeCademy from '../images/codecademy.svg'
import udemy from '../images/udemy-icon.svg'
import wesBos from '../images/wes-bos-icon.png'
import medium from '../images/medium-icon.png'
import devTo from '../images/dev-to.svg'

export const webDevChannels: { name: string, id: string, icon: string }[] = [
    {
        id: 'UCFbNIlppjAuEX4znoulh0Cw',
        name: 'Web Dev Simplified',
        icon: webDevSimplified
    },
    {
        id: 'UClb90NQQcskPUGDIXsQEz5Q',
        name: 'Dev Ed',
        icon: devEd
    },
    {
        id: 'UCW5YeuERMmlnqo4oq8vwUpg',
        name: 'Net Ninja',
        icon: netNinja
    },
    {
        id: 'UCsBjURrPoezykLs9EqgamOA',
        name: 'Fireship',
        icon: fireship
    },
    {
        id: 'UC8butISFwT-Wl7EV0hUK0BQ',
        name: 'Free Code Camp',
        icon: freeCodeCamp
    },
    {
        id: 'UCqrILQNl5Ed9Dz6CGMyvMTQ',
        name: 'Clever Programmer',
        icon: cleverP
    }
]

export const mockArray: {id: string, url: string, title: string, description: string, thumb: string}[] = [
    {
        id: 'someId',
        url: 'https://www.youtube.com/watch?v=Faz6jkFTaa0',
        title: 'title goes here',
        description: 'lorem ipsum',
        thumb: `https://via.placeholder.com/300x200`
    },
    {
        id: 'someId',
        url: 'https://www.youtube.com/watch?v=Faz6jkFTaa0',
        title: 'title goes here',
        description: 'lorem ipsum',
        thumb: `https://via.placeholder.com/300x200`
    }, {
        id: 'someId',
        url: 'https://www.youtube.com/watch?v=Faz6jkFTaa0',
        title: 'title goes here',
        description: 'lorem ipsum',
        thumb: `https://via.placeholder.com/300x200`
    }, {
        id: 'someId',
        url: 'https://www.youtube.com/watch?v=Faz6jkFTaa0',
        title: 'title goes here',
        description: 'lorem ipsum',
        thumb: `https://via.placeholder.com/300x200`
    },
    {
        id: 'someId',
        url: 'https://www.youtube.com/watch?v=Faz6jkFTaa0',
        title: 'title goes here',
        description: 'lorem ipsum',
        thumb: `https://via.placeholder.com/300x200`
    }
]

export const onlineCourses: {platform: string, name: string, url: string, icon: string}[] = [
    {
        platform: 'FreeCodeCamp',
        name: 'Learn React in 1 Hour by Building a Movie Search App',
        url: 'https://www.freecodecamp.org/news/learn-react-in-1-hour-by-building-a-movie-search-app/',
        icon: freeCodeCamp
    },
    {
        platform: 'Udemy',
        name: 'React - The Complete Guide (incl Hooks, React Router, Redux)',
        url: 'https://www.udemy.com/course/react-the-complete-guide-incl-redux/',
        icon: udemy
    },
    {
        platform: 'CodeCademy',
        name: 'Learn React.js: Part I/Part II',
        url: 'https://www.codecademy.com/learn/react-101',
        icon: codeCademy
    },
    {
        platform: 'none',
        name: 'Fullstack Advanced React & GraphQL',
        url: 'https://advancedreact.com/',
        icon: wesBos
    }
]

export const onlineBlogs: {name: string, url: string, icon: string}[] = [
    {
        name: 'Medium',
        url: 'https://medium.com/',
        icon: medium
    },
    {
        name: 'Dev.to',
        url: 'https://dev.to/',
        icon: devTo
    },
    {
        name: 'Sitepoint',
        url: 'https://www.sitepoint.com/',
        icon: sitePoint
    },
    {
        name: 'Smashing Magazine',
        url: 'https://www.smashingmagazine.com/',
        icon: smashingMagazine
    },
    {
        name: 'RobinWieruch',
        url: 'https://www.robinwieruch.de/blog',
        icon: rwieruch
    }
]