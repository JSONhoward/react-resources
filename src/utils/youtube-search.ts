const YOUTUBE_SEARCH_URL = (channel: string, limit: string): string => `https://www.googleapis.com/youtube/v3/search?part=snippet&channelId=${channel}${limit}&q=react&type=video&key=${process.env.REACT_APP_YOUTUBE_API}`

export const getVideoList = async (channel: string, limit = 50): Promise<Video[]> => {
    const response = await fetch(YOUTUBE_SEARCH_URL(channel, `&maxResults=${limit.toString()}`))
    const data = await response.json()

    const videos: Video[] = await data['items'].map((result: { id: { videoId: any }; snippet: { title: any; description: any; thumbnails: { high: { url: any } } } }) => {
        return {
            id: result.id.videoId,
            url: `http://www.youtube.com/watch?v=${result.id.videoId}`,
            title: result.snippet.title,
            description: result.snippet.description,
            thumb: result.snippet.thumbnails.high.url
        }
    })

    return videos
}