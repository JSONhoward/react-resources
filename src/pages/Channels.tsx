import React from 'react'

import Layout from '../components/Layout/Layout';
import ChannelList from '../components/ChannelList/ChannelList';
import VideoPlayer from '../components/VideoPlayer';

const Channels = () => {
    const [portalOpen, setPortalOpen] = React.useState<boolean>(false)
    const [currentVideo, setCurrentVideo] = React.useState<string | null>(null)

    const openPortal = (id: string): void => {
        setPortalOpen(true)
        setCurrentVideo(id)
    }

    const closePortal = (event: React.MouseEvent<HTMLDivElement>): void => {
        const ID = (event.target as HTMLDivElement).id

        if (ID === 'portal') {
            setPortalOpen(false)
            setCurrentVideo(null)
        }
    }

    return (
        <>
            {portalOpen && <VideoPlayer url={currentVideo} close={closePortal} />}
            <Layout>
                <ChannelList openPortal={openPortal} />
            </Layout>
        </>
    )
}

export default Channels
