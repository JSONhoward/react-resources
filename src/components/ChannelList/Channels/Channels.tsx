import React from 'react'
import { FaPlay } from 'react-icons/fa'

import { ChannelIcon, StyledChannels, ChannelName, ChannelVideos, More, VideoImagePlay, VideoImage } from './Channels.styles';
// import { getVideoList } from '../../../utils/youtube-search';

import { Link } from 'react-router-dom';
import { mockArray } from '../../../utils/constants';

type Props = {
    icon: string,
    title: string,
    channelId: string,
    openPortal: (id: string) => void
}

const Channels: React.FC<Props> = ({ icon, title, channelId, openPortal }) => {
    const [videoImages, setVideoImages] = React.useState<Video[] | undefined>(undefined)

    // React.useEffect(() => {
    //     getVideoList(channelId, 3)
    //     .then(videos => setVideoImages(videos))
    //     .catch(error => console.error('Error: ', error))
    // }, [])

    React.useEffect(() => {
        setVideoImages(mockArray)
    }, [])

    const videoImagesList = videoImages?.map((el, i) => {
        return (
            <VideoImage data-testid={`videoImageContainer${i}`} onClick={() => openPortal(el.url)} key={i}>
                <VideoImagePlay>
                    <span>{el.title}</span>
                    <FaPlay size={'3rem'} color={'white'} />
                </VideoImagePlay>
                <img data-testid={`videoImage${i}`} src={el.thumb} alt={el.title} />
            </VideoImage>
        )
    })

    return (
        <StyledChannels>
            <ChannelIcon>
                <ChannelName>
                    {title}
                </ChannelName>
                <a data-testid='channelId' href={`https://www.youtube.com/channel/${channelId}`} target='_blank' rel="noopener noreferrer">
                    <img data-testid='channelIcon' title={title} src={icon} alt={title} />
                </a>
            </ChannelIcon>
            <ChannelVideos>
                {videoImagesList}
                <Link to={`/channel/${channelId}`}>
                    <More>
                        <p>More Videos</p>
                    </More>
                </Link>
            </ChannelVideos>
        </StyledChannels>
    )
}

export default Channels
