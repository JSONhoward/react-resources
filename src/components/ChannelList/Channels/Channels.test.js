import React from 'react'
import renderer from 'react-test-renderer'
import { BrowserRouter as Router } from 'react-router-dom'
import { render, screen, fireEvent } from '@testing-library/react'
import "@testing-library/jest-dom/extend-expect";
import Channels from './Channels';
import { mockArray } from '../../../utils/constants';

describe('snapshot', () => {
    test('snapshot matches', () => {
        const tree = renderer.create(<Router><Channels /></Router>).toJSON()

        expect(tree).toMatchSnapshot()
    })

    test('renders props correctly', () => {
        const testIcon = 'https/via.placeholder.com/150',
            testTitle = 'test title',
            testChannelId = 'testId',
            testOpenPortal = jest.fn(() => console.log('mock portal'))

        render(<Router><Channels icon={testIcon} title={testTitle} channelId={testChannelId} openPortal={testOpenPortal} /></Router>)

        expect(screen.getByText(testTitle)).toBeInTheDocument()
        expect(screen.getByTestId('channelIcon')).toHaveAttribute('src', testIcon)
        expect(screen.getByTestId('channelId')).toHaveAttribute('href', `https://www.youtube.com/channel/${testChannelId}`)
    })
})

describe('renders each channel', () => {
    const testIcon = 'https/via.placeholder.com/150',
        testTitle = 'test title',
        testChannelId = 'testId',
        testOpenPortal = jest.fn(() => console.log('mock portal'))

    render(<Router><Channels icon={testIcon} title={testTitle} channelId={testChannelId} openPortal={testOpenPortal} /></Router>)

    for (let i = 0; i < mockArray.length; i++) {

        expect(screen.getByTestId(`videoImage${i}`)).toHaveAttribute('src', mockArray[i].thumb)
        expect(screen.getByTestId(`videoImage${i}`)).toHaveAttribute('alt', mockArray[i].title)
        fireEvent.click(screen.getByTestId(`videoImageContainer${i}`))
    }
    expect(testOpenPortal).toHaveBeenCalledTimes(mockArray.length)
})