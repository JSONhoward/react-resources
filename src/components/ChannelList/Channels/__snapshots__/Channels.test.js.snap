// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot snapshot matches 1`] = `
<div
  className="sc-bdnylx ieteWa"
>
  <div
    className="sc-dlnjPT fHNJLY"
  >
    <p
      className="sc-hKFyIo erTVFG"
    />
    <a
      data-testid="channelId"
      href="https://www.youtube.com/channel/undefined"
      rel="noopener noreferrer"
      target="_blank"
    >
      <img
        data-testid="channelIcon"
      />
    </a>
  </div>
  <div
    className="sc-eCApGN eFOPNJ"
  >
    <a
      href="/channel/undefined"
      onClick={[Function]}
    >
      <div
        className="sc-iCoHVE fuLZAX"
      >
        <p>
          More Videos
        </p>
      </div>
    </a>
  </div>
</div>
`;
